go/connector/run.go

```diff
@@ -167,29 +204,71 @@ func runCommand(
 	var cmd = exec.Command(args[0], args[1:]...)
 	var fe = new(firstError)
 
-	// Copy |writeLoop| into connector stdin.
-	wc, err := cmd.StdinPipe()
-	if err != nil {
-		return fmt.Errorf("StdinPipe: %w", err)
-	}
-	go func() {
-		defer wc.Close()
-		fe.onError(writeLoop(wc))
-	}()
-
-	var stderrForwarder = ops.NewLogForwardWriter("connector stderr", logrus.InfoLevel, logger)
-
-	// Decode and forward connector stdout to |output|, but intercept a
+	// Decode and forward connector output from socket to |output|, but intercept a
 	// returned error to cancel our context and report through |fe|.
 	// If we didn't cancel, then the connector would run indefinitely.
-	cmd.Stdout = &writeErrInterceptor{
+	var outputInterceptor = &writeErrInterceptor{
 		delegate: output,
 		onError: func(err error) error {
 			fe.onError(err)
 			cancel() // Signal to exit.
 			return err
 		},
 	}
+
+	var conn net.Conn
+	// If port is specified, use TCP socket
+	if port != "" {
+		// Routine dialing a connection to connector
+		go func() {
+			// Try to connect to the connector with a retry mechanism
+			var connectionCounter = 0
+			var err error
+			for {
+				conn, err = net.Dial("tcp", fmt.Sprintf("localhost:%s", port))
+				if err != nil {
+					if connectionCounter > 5 {
+						fe.onError(fmt.Errorf("dialing connection to port %s: %w", port, err))
+						return
+					} else {
+						// Retry after a short wait
+						connectionCounter = connectionCounter + 1
```

Comment:
Will this reliably be enough time, especially in cases where the connector image is being pulled for the first time? I wonder if we could keep a tiny bit of the stdout behavior to allow the connector to send a signal to the runtime on stdin when its listener is ready.

Reply:
@willdonnelly the connector image is pulled before this is invoked, this is invoked as `docker run` is executed.

Reply:
Ah yeah, I see where it is pulled earlier now. Still kind of wondering if this is going to be reliable, but if it isn't it won't be because the image has to be pulled for the first time.

Reply:
@williamhbaker do we want to indefinitely wait for signal from the connector stdout? I think just the fact that a TCP server starts listening on a specific agreed-on port, is a signal. Whether we have a timeout for waiting for ready signal from stdout, or a timeout for connecting to the agreed port, seems the same to me, unless I'm missing something.

Reply:
Good point, we'd need a timeout either way, so having a separate signal wouldn't accomplish anything. Thanks!